1/TransactionBuilder.js

```diff
@@ -77,17 +94,42 @@ class TransactionBuilder {
      */
     input(fs, ids) {
 
+        if ( this._signatures !== undefined ) {
+            throw new Error("Attempting to add new input, expecting signatures only")
+        }
+
         //if this is setup as coinbase, prevent additional inputs
         if (Object.keys(this._inputs).find(address => address === constant.COINBASE_ADDRESS_PUBLIC)) throw new Error('Cannot add an additional input to a coinbase transaction');
 
-        if (!fctAddressUtil.isValidPrivateAddress(fs)) throw new Error("Input address must be a valid private Factoid address");
         if (!util.validateNFIds(ids)) throw new Error("Invalid ID range: " + JSON.stringify(ids));
 
-        //check that outputs does not contain this address
-        if (this._outputs[fctAddressUtil.getPublicAddress(fs)]) throw new Error("Input address already occurs in outputs");
+        //if it isn't a private address and instead a public address then, the fs should be a public key      
+        if (fctAddressUtil.isValidPrivateAddress(fs)) { //first check to see if valid private address
+
+            //check that outputs does not contain this address
+            if (this._outputs[fctAddressUtil.getPublicAddress(fs)]) throw new Error("Input address already occurs in outputs");
+
+            this._keys.push(nacl.keyPair.fromSeed(fctAddressUtil.addressToKey(fs)));
+            this._inputs[fctAddressUtil.getPublicAddress(fs)] = ids;
+
+        } else {
+
+            if ( fctAddressUtil.isValidPublicFctAddress(fs) ) { //check to see if user passed in a public fct address
```

Comment:
this is rather unexpected for the end user to have to provide a raw public key here. Couldn't it work with a public FCT address? If would be more symmetrical to the case you provide FCT private address (and not private key). When the user call `pkSignature` you can convert the PK to an FCT public address and search for its present in the inputs to do the same validation you are doing now.

Reply:
Good suggestion.

Reply:
this is done.